core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -102,27 +110,52 @@ public void execute() throws Exception {
         }
 
         totalShards.incrementAndGet();
-        pendingShards.incrementAndGet();
+        pendingActions.incrementAndGet();
         primaryResult = primary.perform(request);
         final ReplicaRequest replicaRequest = primaryResult.replicaRequest();
         assert replicaRequest.primaryTerm() > 0 : "replicaRequest doesn't have a primary term";
         if (logger.isTraceEnabled()) {
             logger.trace("[{}] op [{}] completed on primary for request [{}]", primaryId, opType, request);
         }
 
-        performOnReplicas(primaryId, replicaRequest);
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        ClusterState clusterState = clusterStateSupplier.get();
+        final List<ShardRouting> shards = getShards(primaryId, clusterState);
+
+        if (markUnavailableShardsAsStale) {
+            Set<String> inSyncAllocationIds = getInSyncAllocationIds(primaryId, clusterState);
+
+            if (inSyncAllocationIds.isEmpty() == false && shards.isEmpty() == false) {
+                Set<String> availableAllocationIds = shards.stream()
+                    .map(ShardRouting::allocationId)
+                    .filter(Objects::nonNull)
+                    .map(AllocationId::getId)
+                    .collect(Collectors.toSet());
+
+                // if inSyncAllocationIds contains allocation ids of shards that don't exist in RoutingTable, mark copies as stale
+                for (String allocationId : Sets.difference(inSyncAllocationIds, availableAllocationIds)) {
+                    // mark copy as stale
+                    pendingActions.incrementAndGet();
+                    replicasProxy.markShardCopyAsStale(primaryId, allocationId, replicaRequest.primaryTerm(),
+                        ReplicationOperation.this::decPendingAndFinishIfNeeded,
+                        ReplicationOperation.this::onPrimaryDemoted,
+                        throwable -> decPendingAndFinishIfNeeded()
+                    );
+                }
+            }
+        }
+
+        performOnReplicas(replicaRequest, shards);
```

Comment:
Actually, I believe that's the purpose of `pendingActions` - we don't respond to the replication request until `pendingActions == 0`